### ISSUE TITLE
control_dependencies with assert_equal

### ISSUE THREAD

**Issue by bhack** [tensorflow/tensorflow#39305]
Thank you for submitting a TensorFlow documentation issue. Per our GitHub
policy, we only address code/doc bugs, performance issues, feature requests, and
build/installation issues on GitHub.

The TensorFlow docs are open source! To get involved, read the documentation
contributor guide: https://www.tensorflow.org/community/contribute/docs

## URL(s) with the issue:

Please provide a link to the documentation entry, for example:
https://www.tensorflow.org/api_docs/python/tf/control_dependencies

## Description of issue (what needs changing):
Do we need to mention the debug use case in https://www.tensorflow.org/api_docs/python/tf/debugging/assert_equal#returns ?
### Clear description
We declare in the note
> Note: In TensorFlow 2 with eager and/or Autograph, you should not require this method, as code executes in the expected order. Only use tf.control_dependencies when working with v1-style code or in a graph context such as inside Dataset.map.
But there is any direct reference to the `assert_equal` use case

For example, why should someone use this method? How is it useful?
Take a look at the issue [here](https://github.com/tensorflow/addons/issues/1794)

**Comment by rmothukuru** [comment]
@bhack,
Sorry for the delayed response. Can you please elaborate on your point? Thanks!

**Comment by bhack** [comment]
Check https://github.com/tensorflow/tensorflow/issues/48699#issuecomment-825653937 /cc @mdanatg @MarkDaoust

**Comment by sachinprasadhs** [comment]
Hi, The documentation has been updated [here](https://www.tensorflow.org/api_docs/python/tf/control_dependencies?version=nightly) with the below details which states:
`In TensorFlow 2 with eager and/or Autograph, you should not need this method most of the times, as ops execute in the expected order thanks to automatic control dependencies. Only use it to manually control ordering, for example as a workaround to known issues such as [tf.function](https://www.tensorflow.org/api_docs/python/tf/function) with tf.debugging.assert* and [tf.py_function](https://www.tensorflow.org/api_docs/python/tf/py_function).`
Could you please verify and close the issue. Thanks!

**Comment by mdanatg** [comment]
I think assert ops have a special requirement that (at least until a few months ago) defeated the usual side effect analysis, because they need to guard either stateless operations or stateful ops that aren't in the same class as the assert.

For example:

```
tf.assert_not_equal(x, 0)
y = tf.math.divide(1, x)
return y
```

Under naive rules, `tf.math.divide` is stateless and so it may proceed in parallel with `assert_not_equal`. Meaning that it can well execute before the assert blows up, and you get the divide by zero instead. I'm not sure, has this been fixed in the meantime?

Specifically, what naive rules would create:
```
# after acd:
op = tf.assert_not_equal(x, 0)
y = tf.math.divide(1, x)
with tf.control_dependencies([op]):  # add control deps on all stateful ops that aren't yet guarded
  ret_y = tf.identity(y)
return ret_y
```

The correct rules would instead do:

```
# better:
op = tf.assert_not_equal(x, 0)
with tf.control_dependencies([op]):  # add control deps on all asserts that share inputs with the op (i.e. x)
  y = tf.math.divide(1, x)
# all stateful ops guarded at this point, nothing further needed
return y
```

This issue is somewhat masked at present because most stateful ops are considered in the same class as assert, which is a separate issue in itself.

**Comment by bhack** [comment]
```python
import tensorflow as tf
@tf.function(jit_compile=True)
def test(x):
  tf.debugging.assert_none_equal(x, 0)
  y = tf.math.divide(1, x)
  return y
print(test(1))
test(0)
```

I see
```python
def tf__test(x):
    with ag__.FunctionScope('test', 'fscope', ag__.ConversionOptions(recursive=True, user_requested=True, optional_features=(), internal_convert_user_code=True)) as fscope:
        do_return = False
        retval_ = ag__.UndefinedReturnValue()
        ag__.converted_call(ag__.ld(tf).debugging.assert_none_equal, (ag__.ld(x), 0), None, fscope)
        y = ag__.converted_call(ag__.ld(tf).math.divide, (1, ag__.ld(x)), None, fscope)
        try:
            do_return = True
            retval_ = ag__.ld(y)
        except:
            do_return = False
            raise
        return fscope.ret(retval_, do_return)
```

**Comment by mdanatg** [comment]
[ACD](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/framework/auto_control_deps.py) is separate from AutoGraph (long story why). Anyway, ACD runs post AutoGraph. If you look at the function's graph, things should become more revealing: `print(test.get_concrete_function(1).graph.as_graph_def())`

Side note: you may want to first [convert that graph to MLIR](https://www.tensorflow.org/api_docs/python/tf/mlir/experimental/convert_graph_def) and look at that instead - it should be more readable.

**Comment by bhack** [comment]
```mlir
module attributes {tf.versions = {bad_consumers = [], min_consumer = 0 : i32, producer = 1331 : i32}} {
  func.func @main() {
    tf_executor.graph {
      %outputs, %control = tf_executor.island wraps "tf.Const"() {device = "", value = dense<"Condition x != y did not hold element-wise:"> : tensor<!tf_type.string>} : () -> tensor<!tf_type.string>
      %outputs_0, %control_1 = tf_executor.island wraps "tf.Const"() {device = "", value = dense<"x (assert_none_equal/x:0) = "> : tensor<!tf_type.string>} : () -> tensor<!tf_type.string>
      %outputs_2, %control_3 = tf_executor.island wraps "tf.Const"() {device = "", value = dense<"y (assert_none_equal/y:0) = "> : tensor<!tf_type.string>} : () -> tensor<!tf_type.string>
      tf_executor.fetch
    }
    return
  }
}
```

**Comment by mdanatg** [comment]
Ah, looks like you need some special parameters to get the right MLIR output - I don't know what they are :s. So scratch that last, better to look at the GraphDef...

**Comment by bhack** [comment]
```python
node {
  name: "assert_none_equal/x"
  op: "Const"
  attr {
    key: "dtype"
    value {
      type: DT_INT32
    }
  }
  attr {
    key: "value"
    value {
      tensor {
        dtype: DT_INT32
        tensor_shape {
        }
        int_val: 1
      }
    }
  }
}
node {
  name: "assert_none_equal/y"
  op: "Const"
  attr {
    key: "dtype"
    value {
      type: DT_INT32
    }
  }
  attr {
    key: "value"
    value {
      tensor {
        dtype: DT_INT32
        tensor_shape {
        }
        int_val: 0
      }
    }
  }
}
node {
  name: "assert_none_equal/NotEqual"
  op: "NotEqual"
  input: "assert_none_equal/x"
  input: "assert_none_equal/y"
  attr {
    key: "T"
    value {
      type: DT_INT32
    }
  }
  attr {
    key: "incompatible_shape_error"
    value {
      b: true
    }
  }
}
node {
  name: "assert_none_equal/Rank"
  op: "Const"
  attr {
    key: "dtype"
    value {
      type: DT_INT32
    }
  }
  attr {
    key: "value"
    value {
      tensor {
        dtype: DT_INT32
        tensor_shape {
        }
        int_val: 0
      }
    }
  }
}
node {
  name: "assert_none_equal/range/start"
  op: "Const"
  attr {
    key: "dtype"
    value {
      type: DT_INT32
    }
  }
  attr {
    key: "value"
    value {
      tensor {
        dtype: DT_INT32
        tensor_shape {
        }
        int_val: 0
      }
    }
  }
}
node {
  name: "assert_none_equal/range/delta"
  op: "Const"
  attr {
    key: "dtype"
    value {
      type: DT_INT32
    }
  }
  attr {
    key: "value"
    value {
      tensor {
        dtype: DT_INT32
        tensor_shape {
        }
        int_val: 1
      }
    }
  }
}
node {
  name: "assert_none_equal/range"
  op: "Range"
  input: "assert_none_equal/range/start"
  input: "assert_none_equal/Rank"
  input: "assert_none_equal/range/delta"
  attr {
    key: "Tidx"
    value {
      type: DT_INT32
    }
  }
}
node {
  name: "assert_none_equal/All"
  op: "All"
  input: "assert_none_equal/NotEqual"
  input: "assert_none_equal/range"
  attr {
    key: "Tidx"
    value {
      type: DT_INT32
    }
  }
  attr {
    key: "keep_dims"
    value {
      b: false
    }
  }
}
node {
  name: "assert_none_equal/Assert/Const"
  op: "Const"
  attr {
    key: "dtype"
    value {
      type: DT_STRING
    }
  }
  attr {
    key: "value"
    value {
      tensor {
        dtype: DT_STRING
        tensor_shape {
        }
        string_val: "Condition x != y did not hold element-wise:"
      }
    }
  }
}
node {
  name: "assert_none_equal/Assert/Const_1"
  op: "Const"
  attr {
    key: "dtype"
    value {
      type: DT_STRING
    }
  }
  attr {
    key: "value"
    value {
      tensor {
        dtype: DT_STRING
        tensor_shape {
        }
        string_val: "x (assert_none_equal/x:0) = "
      }
    }
  }
}
node {
  name: "assert_none_equal/Assert/Const_2"
  op: "Const"
  attr {
    key: "dtype"
    value {
      type: DT_STRING
    }
  }
  attr {
    key: "value"
    value {
      tensor {
        dtype: DT_STRING
        tensor_shape {
        }
        string_val: "y (assert_none_equal/y:0) = "
      }
    }
  }
}
node {
  name: "assert_none_equal/Assert/Assert/data_0"
  op: "Const"
  attr {
    key: "dtype"
    value {
      type: DT_STRING
    }
  }
  attr {
    key: "value"
    value {
      tensor {
        dtype: DT_STRING
        tensor_shape {
        }
        string_val: "Condition x != y did not hold element-wise:"
      }
    }
  }
}
node {
  name: "assert_none_equal/Assert/Assert/data_1"
  op: "Const"
  attr {
    key: "dtype"
    value {
      type: DT_STRING
    }
  }
  attr {
    key: "value"
    value {
      tensor {
        dtype: DT_STRING
        tensor_shape {
        }
        string_val: "x (assert_none_equal/x:0) = "
      }
    }
  }
}
node {
  name: "assert_none_equal/Assert/Assert/data_3"
  op: "Const"
  attr {
    key: "dtype"
    value {
      type: DT_STRING
    }
  }
  attr {
    key: "value"
    value {
      tensor {
        dtype: DT_STRING
        tensor_shape {
        }
        string_val: "y (assert_none_equal/y:0) = "
      }
    }
  }
}
node {
  name: "assert_none_equal/Assert/Assert"
  op: "Assert"
  input: "assert_none_equal/All"
  input: "assert_none_equal/Assert/Assert/data_0"
  input: "assert_none_equal/Assert/Assert/data_1"
  input: "assert_none_equal/x"
  input: "assert_none_equal/Assert/Assert/data_3"
  input: "assert_none_equal/y"
  attr {
    key: "T"
    value {
      list {
        type: DT_STRING
        type: DT_STRING
        type: DT_INT32
        type: DT_STRING
        type: DT_INT32
      }
    }
  }
  attr {
    key: "summarize"
    value {
      i: 3
    }
  }
}
node {
  name: "Const"
  op: "Const"
  attr {
    key: "dtype"
    value {
      type: DT_INT32
    }
  }
  attr {
    key: "value"
    value {
      tensor {
        dtype: DT_INT32
        tensor_shape {
        }
        int_val: 1
      }
    }
  }
}
node {
  name: "truediv/y"
  op: "Const"
  attr {
    key: "dtype"
    value {
      type: DT_INT32
    }
  }
  attr {
    key: "value"
    value {
      tensor {
        dtype: DT_INT32
        tensor_shape {
        }
        int_val: 1
      }
    }
  }
}
node {
  name: "truediv/Cast"
  op: "Cast"
  input: "Const"
  attr {
    key: "DstT"
    value {
      type: DT_DOUBLE
    }
  }
  attr {
    key: "SrcT"
    value {
      type: DT_INT32
    }
  }
  attr {
    key: "Truncate"
    value {
      b: false
    }
  }
}
node {
  name: "truediv/Cast_1"
  op: "Cast"
  input: "truediv/y"
  attr {
    key: "DstT"
    value {
      type: DT_DOUBLE
    }
  }
  attr {
    key: "SrcT"
    value {
      type: DT_INT32
    }
  }
  attr {
    key: "Truncate"
    value {
      b: false
    }
  }
}
node {
  name: "truediv"
  op: "RealDiv"
  input: "truediv/Cast"
  input: "truediv/Cast_1"
  attr {
    key: "T"
    value {
      type: DT_DOUBLE
    }
  }
}
node {
  name: "Identity"
  op: "Identity"
  input: "truediv"
  input: "^NoOp"
  attr {
    key: "T"
    value {
      type: DT_DOUBLE
    }
  }
}
node {
  name: "NoOp"
  op: "NoOp"
  input: "^assert_none_equal/Assert/Assert"
  attr {
    key: "_acd_function_control_output"
    value {
      b: true
    }
  }
}
versions {
  producer: 1331
}
```

**Comment by mdanatg** [comment]
Right, so if you scroll past the boilerplate, you can see at the bottom that the final `NoOp` (that's basically going to be tied to the return value, as you can see from its attribute `key: "_acd_function_control_output"`) has a control dependency on the assert op (`input: "^assert_none_equal/Assert/Assert"`). Whereas the division node (`name: "truediv"`) has no control depdendencies! So in real life, you either get an assertion failure or a divide by zero, at random. For things to work as expected, it's the division op which should have the control dep.

**Comment by bhack** [comment]
A bit cumbersome to debug but ok 😁

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
